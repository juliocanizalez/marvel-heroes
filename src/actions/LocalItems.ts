import { ILocalItem } from '../reducers/localItems/ILocalItemsState';
import {
  IAddBookmark,
  IBookmarksReset,
  IHiddenItemsReset,
  IHideLocalItem,
  ILocalItemsReset,
  IRemoveBookmark,
} from '../reducers/localItems/LocalItemsActions';

const hideLocalItem = (item: ILocalItem): IHideLocalItem => ({
  type: 'HIDE_LOCAL_ITEM',
  payload: {
    item,
  },
});

const addBookmark = (item: ILocalItem): IAddBookmark => ({
  type: 'ADD_BOOKMARK',
  payload: {
    item,
  },
});

const removeBookmark = (item: ILocalItem): IRemoveBookmark => ({
  type: 'REMOVE_BOOKMARK',
  payload: {
    item,
  },
});

const resetBookmarks = (): IBookmarksReset => ({
  type: 'RESET_BOOKMARKS',
});

const resetHiddenItems = (): IHiddenItemsReset => ({
  type: 'RESET_HIDDEN_ITEMS',
});

const resetLocalItems = (): ILocalItemsReset => ({
  type: 'RESET_LOCAL_ITEMS',
});

export {
  hideLocalItem,
  addBookmark,
  removeBookmark,
  resetBookmarks,
  resetHiddenItems,
  resetLocalItems,
};
