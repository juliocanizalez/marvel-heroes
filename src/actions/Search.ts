import ICharacter from '../interfaces/ICharacter';
import IGenericApiResponse from '../interfaces/IGenericApiResponse';
import {
  ISetOffset,
  ISetBaseUrl,
  IReset,
  ISetAllParams,
  ISetName,
  ISetComic,
  ISetStory,
  ISetLoading,
  ISetError,
  ISetData,
  ISetAsyncContent,
} from '../reducers/search/SearchActions';

const setBaseUrl = (baseUrl: string): ISetBaseUrl => ({
  type: 'SET_BASE_URL',
  payload: {
    baseUrl,
  },
});

const setOffset = (offset: number): ISetOffset => ({
  type: 'SET_OFFSET',
  payload: {
    offset,
  },
});

const setName = (name: string): ISetName => ({
  type: 'SET_NAME',
  payload: {
    name,
  },
});

const setComic = (comic: string): ISetComic => ({
  type: 'SET_COMIC',
  payload: {
    comic,
  },
});

const setStory = (story: string): ISetStory => ({
  type: 'SET_STORY',
  payload: {
    story,
  },
});

const setAllParams = (
  offset: number,
  name: string,
  comic: string,
  story: string,
): ISetAllParams => ({
  type: 'SET_ALL_PARAMS',
  payload: {
    offset,
    name,
    comic,
    story,
  },
});

const setLoading = (loading: boolean): ISetLoading => ({
  type: 'SET_LOADING',
  payload: {
    loading,
  },
});

const setError = (error: string): ISetError => ({
  type: 'SET_ERROR',
  payload: {
    error,
  },
});

const setData = (data: IGenericApiResponse<ICharacter> | null): ISetData => ({
  type: 'SET_DATA',
  payload: {
    data,
  },
});

const setAsyncContent = (
  loading: boolean,
  error: string,
  data: IGenericApiResponse<ICharacter> | null,
): ISetAsyncContent => ({
  type: 'SET_ASYNC_CONTENT',
  payload: {
    loading,
    error,
    data,
  },
});

const reset = (): IReset => ({
  type: 'RESET',
});

export {
  setBaseUrl,
  setOffset,
  setName,
  setComic,
  setStory,
  setAllParams,
  setLoading,
  setError,
  setData,
  setAsyncContent,
  reset,
};
