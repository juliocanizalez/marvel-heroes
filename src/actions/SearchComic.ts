import IComic from '../interfaces/IComic';
import IGenericApiResponse from '../interfaces/IGenericApiResponse';
import {
  IComicReset,
  IComicSetAllParams,
  IComicSetAsyncContent,
  IComicSetBaseUrl,
  IComicSetData,
  IComicSetError,
  IComicSetFormat,
  IComicSetLoading,
  IComicSetTitle,
} from '../reducers/searchComic/SearchComicActions';

const setComicBaseUrl = (baseUrl: string): IComicSetBaseUrl => ({
  type: 'SET_BASE_URL_COMIC',
  payload: {
    baseUrl,
  },
});

const setComicTitle = (title: string): IComicSetTitle => ({
  type: 'SET_TITLE_COMIC',
  payload: {
    title,
  },
});

const setComicFormat = (format: string): IComicSetFormat => ({
  type: 'SET_FORMAT_COMIC',
  payload: {
    format,
  },
});

const setComicAllParams = (offset: number, format: string, title: string): IComicSetAllParams => ({
  type: 'SET_ALL_PARAMS_COMIC',
  payload: {
    offset,
    format,
    title,
  },
});

const setComicLoading = (loading: boolean): IComicSetLoading => ({
  type: 'SET_LOADING_COMIC',
  payload: {
    loading,
  },
});

const setComicError = (error: string): IComicSetError => ({
  type: 'SET_ERROR_COMIC',
  payload: {
    error,
  },
});

const setComicData = (data: IGenericApiResponse<IComic> | null): IComicSetData => ({
  type: 'SET_DATA_COMIC',
  payload: {
    data,
  },
});

const setComicAsyncContent = (
  loading: boolean,
  error: string,
  data: IGenericApiResponse<IComic> | null,
): IComicSetAsyncContent => ({
  type: 'SET_ASYNC_CONTENT_COMIC',
  payload: {
    loading,
    error,
    data,
  },
});

const comicReset = (): IComicReset => ({
  type: 'RESET_COMIC',
});

export {
  setComicBaseUrl,
  setComicTitle,
  setComicFormat,
  setComicAllParams,
  setComicLoading,
  setComicError,
  setComicData,
  setComicAsyncContent,
  comicReset,
};
