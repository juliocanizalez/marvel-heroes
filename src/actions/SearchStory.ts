import IStory from '../interfaces/IStory';
import IGenericApiResponse from '../interfaces/IGenericApiResponse';
import {
  IStoryReset,
  IStorySetAllParams,
  IStorySetAsyncContent,
  IStorySetBaseUrl,
  IStorySetData,
  IStorySetError,
  IStorySetLoading,
  IStorySetTitle,
} from '../reducers/searchStory/SearchStoryActions';

const setStoryBaseUrl = (baseUrl: string): IStorySetBaseUrl => ({
  type: 'SET_BASE_URL_STORY',
  payload: {
    baseUrl,
  },
});

const setStoryTitle = (title: string): IStorySetTitle => ({
  type: 'SET_TITLE_STORY',
  payload: {
    title,
  },
});

const setStoryAllParams = (offset: number, title: string): IStorySetAllParams => ({
  type: 'SET_ALL_PARAMS_STORY',
  payload: {
    offset,
    title,
  },
});

const setStoryLoading = (loading: boolean): IStorySetLoading => ({
  type: 'SET_LOADING_STORY',
  payload: {
    loading,
  },
});

const setStoryError = (error: string): IStorySetError => ({
  type: 'SET_ERROR_STORY',
  payload: {
    error,
  },
});

const setStoryData = (data: IGenericApiResponse<IStory> | null): IStorySetData => ({
  type: 'SET_DATA_STORY',
  payload: {
    data,
  },
});

const setStoryAsyncContent = (
  loading: boolean,
  error: string,
  data: IGenericApiResponse<IStory> | null,
): IStorySetAsyncContent => ({
  type: 'SET_ASYNC_CONTENT_STORY',
  payload: {
    loading,
    error,
    data,
  },
});

const storyReset = (): IStoryReset => ({
  type: 'RESET_STORY',
});

export {
  setStoryBaseUrl,
  setStoryTitle,
  setStoryAllParams,
  setStoryLoading,
  setStoryError,
  setStoryData,
  setStoryAsyncContent,
  storyReset,
};
