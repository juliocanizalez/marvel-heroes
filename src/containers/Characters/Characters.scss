@use '../../styles/global';

.main-content{
  text-align: center;
  @extend %marvel-font;
}

.back-button {
  color: #fff;
  outline: none;
  cursor: pointer;
  background-color: global.$primary;
  height: 2rem;
  padding: 0.5rem 1rem;
  border: none;
}

.action-btn{
  color: #fff;
  outline: none;
  cursor: pointer;
  background-color: global.$primary;
  height: 2rem;
  padding: 0.5rem 1rem;
  border: none;
  margin-right: 1rem;
}

.search-header__list{
  display: flex;
  justify-content: space-around;
  margin-top: 1rem;
}

.search-title{
  margin-top: 1.4rem;
  font-weight: global.$font-bold;
  font-size: 1.5rem;
}

.page-title{
  h1{
    font-size: 2.5rem;
  }
}

.search-select{
  display: flex;
  justify-content: space-around;
  margin-bottom: 2rem;
}

.cards-content{
    display: grid;
    gap: 1rem;
    grid-template-columns: repeat(5, 1fr);
  }