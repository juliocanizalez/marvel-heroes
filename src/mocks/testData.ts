/* eslint-disable */
import ICharacter from '../interfaces/ICharacter';
import IComic from '../interfaces/IComic';
import IStory from '../interfaces/IStory';

const createGenericApiResponse = <T>(results: T[]) => ({
  data: {
    offset: 0,
    limit: 12,
    total: 1,
    count: 1,
    results,
  },
});

const emptyResponse = () => ({
  data: {
    offset: 0,
    limit: 12,
    total: 0,
    count: 0,
    results: [],
  },
});

const charactersResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      comics: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
    },
  ]);

const characterResponse = () =>
  createGenericApiResponse<ICharacter>([
    {
      id: 1,
      name: 'spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
    },
  ]);

const comicsResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [],
      characters: {} as any,
      stories: {} as any,
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const comicResponse = () =>
  createGenericApiResponse<IComic>([
    {
      id: 1,
      title: 'comic spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      urls: [
        {
          type: 'detail',
          url: 'http://marvel.com/characters/2854/adam_warlock?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
        {
          type: 'wiki',
          url: 'http://marvel.com/universe/Warlock,_Adam?utm_campaign=apiRef&utm_source=10dea460449717d1dce670f173fe30b9',
        },
      ],
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      stories: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Story',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      variants: {} as any,
      collections: {} as any,
      collectedIssues: {} as any,
      creators: {} as any,
      digitalId: 0,
      issueNumber: 0,
      variantDescription: '',
      isbn: '',
      upc: '',
      diamondCode: '',
      ean: '',
      issn: '',
      format: '',
      pageCount: 10,
      textObjects: [],
      dates: [],
      prices: [],
      images: [],
    },
  ]);

const storiesResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {} as any,
      characters: {} as any,
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {} as any,
      type: '',
    },
  ]);

const storyResponse = () =>
  createGenericApiResponse<IStory>([
    {
      id: 1,
      title: 'story spiderman',
      description: 'description',
      thumbnail: {
        path: 'path',
        extension: 'jpg',
      },
      modified: new Date(),
      resourceURI: '',
      comics: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Spiderman Comic',
          },
        ],
      },
      characters: {
        available: 1,
        returned: 1,
        collectionURI: '',
        items: [
          {
            resourceURI: '',
            name: 'Char Spiderman',
          },
        ],
      },
      events: {} as any,
      series: {} as any,
      creators: {} as any,
      originalIssue: {
        resourceURI: 'apiUrl/comics/1',
        name: 'Spiderman Comic',
      },
      type: '',
    },
  ]);

export {
  createGenericApiResponse,
  emptyResponse,
  charactersResponse,
  characterResponse,
  comicsResponse,
  comicResponse,
  storiesResponse,
  storyResponse,
};
