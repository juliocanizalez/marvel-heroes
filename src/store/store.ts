/* eslint-disable no-underscore-dangle */
import { createStore, combineReducers, StoreEnhancer } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import SearchReducer from '../reducers/search/SearchReducer';
import SearchComicReducer from '../reducers/searchComic/SearchComicReducer';
import ViewItemReducer from '../reducers/viewItem/ViewItemReducer';
import LocalItemsReducer from '../reducers/localItems/LocalItemsReducer';

type WindowWithDevTools = Window & {
  __REDUX_DEVTOOLS_EXTENSION__: () => StoreEnhancer<unknown, {}>;
};

const isReduxDevtoolsExtenstionExist = (
  arg: Window | WindowWithDevTools,
): arg is WindowWithDevTools => '__REDUX_DEVTOOLS_EXTENSION__' in arg;

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['localItems'],
};

const rootReducer = combineReducers({
  search: SearchReducer,
  searchComic: SearchComicReducer,
  viewItem: ViewItemReducer,
  localItems: LocalItemsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = createStore(
  persistedReducer,
  isReduxDevtoolsExtenstionExist(window) ? window.__REDUX_DEVTOOLS_EXTENSION__() : undefined,
);

export const persistor = persistStore(store);

export type IRootState = ReturnType<typeof rootReducer>;

export default store;
