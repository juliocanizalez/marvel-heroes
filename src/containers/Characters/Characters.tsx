/* eslint-disable */
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import queryString from 'query-string';
import { useHistory } from 'react-router-dom';
import { debounce } from 'lodash';
import { faBookmark as faBookmarkSolid } from '@fortawesome/free-solid-svg-icons';
import { faBookmark as faBookmarkRegular } from '@fortawesome/free-regular-svg-icons';

import { IRootState } from '../../store/store';
import IGenericApiResponse from '../../interfaces/IGenericApiResponse';
import ICharacter from '../../interfaces/ICharacter';
import IComic from '../../interfaces/IComic';
import IStory from '../../interfaces/IStory';
import { getComicsByOffsetLimit, getStoriesByOffsetLimit } from '../../helpers/FetchService';
import Select from '../../components/Select/Select';
import Card from '../../components/Card/Card';
import Loading from '../../components/Loading/Loading';
import Pagination from '../../components/Pagination/Pagination';
import { getQueryVariable, imagePlaceHolder } from '../../utils/utils';
import { get } from '../../api/FetchInfo';
import { addBookmark, hideLocalItem, removeBookmark } from '../../actions/LocalItems';
import {
  reset,
  setAllParams,
  setAsyncContent,
  setBaseUrl,
  setComic,
  setName,
  setStory,
} from '../../actions/Search';
import './Characters.scss';

const ListCharactersPage = () => {
  enum QuerysParams {
    Page = 'page',
    Name = 'name',
    Comic = 'comic',
    Story = 'story',
  }

  const isMounted = useRef(true);
  const history = useHistory();
  const dispatch = useDispatch();
  const [comics, setComics] = useState<IComic[]>([]);
  const [stories, setStories] = useState<IStory[]>([]);
  const {
    url,
    limit,
    name,
    comic,
    story,
    loading,
    error,
    data: genericResponse,
  } = useSelector((state: IRootState) => state.search);
  const { hiddenItems, bookmarks } = useSelector((state: IRootState) => state.localItems);

  const { data } = genericResponse ?? {};
  const { results } = data ?? {};
  const [characters, setCharacters] = useState<ICharacter[] | null>(null);

  const changeUrlParams = (
    newPage: number,
    newName: string,
    newComic: string,
    newStory: string,
  ) => {
    history.push(
      `?${queryString.stringify({
        page: newPage,
        story: newStory === '' ? undefined : newStory,
        comic: newComic === '' ? undefined : newComic,
        name: newName === '' ? undefined : newName,
      })}`,
    );
  };

  const handleChangePage = (newPage: number) => {
    window.scrollTo({ top: 300 });
    changeUrlParams(newPage, name, comic, story);
  };

  const debouncedNameChange = useCallback(
    debounce(
      (name: string, comic: string, story: string) => changeUrlParams(1, name, comic, story),
      1000,
    ),
    [],
  );

  const handleNameChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setName(target.value));
    debouncedNameChange(target.value, comic, story);
  };

  const handleComicChange = ({ target }: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(setComic(target.value));

    changeUrlParams(1, name, target.value, story);
  };

  const handleStoryChange = ({ target }: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(setStory(target.value));

    changeUrlParams(1, name, comic, target.value);
  };

  const handleSubmit = (e: React.ChangeEvent<HTMLFormElement>) => e.preventDefault();

  const handleViewMore = (id: number) => history.push(`characters/${id}`);

  const handleHideItem = (id: number) => dispatch(hideLocalItem({ id, type: 'CHARACTER' }));

  const handleAddBookmark = (id: number) => dispatch(addBookmark({ id, type: 'CHARACTER' }));

  const handleRemoveBookmark = (id: number) => dispatch(removeBookmark({ id, type: 'CHARACTER' }));

  useEffect(() => {
    dispatch(setBaseUrl(`${process.env.REACT_APP_API_URL}v1/public/characters`));

    getComicsByOffsetLimit(0, 10)
      .then((genRes) => {
        if (isMounted.current) {
          if (genRes) {
            const { data } = genRes;
            const { results } = data;
            setComics(results);
          } else {
            setComics([]);
          }
        }
      })
      .catch((e) => {
        if (isMounted.current) {
          setComics([]);
        }
      });

    getStoriesByOffsetLimit(40, 10)
      .then((genRes) => {
        if (isMounted.current) {
          if (genRes) {
            const { data } = genRes;
            const { results } = data;
            setStories(results);
          } else {
            setStories([]);
          }
        }
      })
      .catch((e) => {
        if (isMounted.current) {
          setStories([]);
        }
      });

    return () => {
      isMounted.current = false;
      dispatch(reset());
    };
  }, []);

  useEffect(() => {
    const newPage = getQueryVariable(QuerysParams.Page) ? +getQueryVariable(QuerysParams.Page) : 1;

    const newName = getQueryVariable(QuerysParams.Name)
      ? decodeURI(getQueryVariable(QuerysParams.Name))
      : '';

    const newComic = getQueryVariable(QuerysParams.Comic)
      ? decodeURI(getQueryVariable(QuerysParams.Comic))
      : '';

    const newStory = getQueryVariable(QuerysParams.Story)
      ? decodeURI(getQueryVariable(QuerysParams.Story))
      : '';

    const newOffset = Math.ceil((newPage - 1) * limit);

    dispatch(setAllParams(newOffset, newName, newComic, newStory));
  }, [history.location, limit]);

  useEffect(() => {
    if (url) {
      dispatch(setAsyncContent(true, '', null));
      get<IGenericApiResponse<ICharacter>>(url)
        .then((res) => {
          dispatch(setAsyncContent(false, '', res));
        })
        .catch((err) => {
          dispatch(setAsyncContent(false, 'Could not load characters', null));
        });
    }
  }, [url]);

  useEffect(() => {
    if (results) {
      const hiddenChars = hiddenItems.filter((item) => item.type === 'CHARACTER');
      setCharacters(
        results.filter((char) => {
          const hiddenCharFilter = hiddenChars.filter((item) => item.id === char.id);
          if (hiddenCharFilter.length > 0) {
            return false;
          }
          return true;
        }),
      );
    }
  }, [results, hiddenItems]);

  return (
    <div className='main-content '>
      <div className='page-title'>
        <h1>Characters</h1>
      </div>
      <div className='search-filters-form mb-5'>
        <form onSubmit={handleSubmit}>
          <div className='search-title'>Search your hero</div>
          <div className='search-header__list'>
            <h3>Name</h3>
            <h3>Comic</h3>
            <h3>Story</h3>
          </div>

          <div className='search-select'>
            <div className='search-value'>
              <input
                type='text'
                name='name'
                placeholder='Hero¬¥s name'
                autoComplete='off'
                className='search-input'
                value={name}
                onChange={handleNameChange}
              />
            </div>

            <div className='search-value'>
              <Select
                defaultOptionText='Select a comic'
                items={comics}
                value={comic}
                onChange={handleComicChange}
              />
            </div>

            <div className='search-value'>
              <Select
                defaultOptionText='Select a story'
                items={stories}
                value={story}
                onChange={handleStoryChange}
              />
            </div>
          </div>
        </form>
      </div>
      {loading && <Loading />}
      {error && <h2 className='error-message'>Could not load characters üòì</h2>}
      <div className='cards'>
        <div className='cards-content'>
          {!loading &&
            characters &&
            characters.map((char) => {
              const inBookmark = bookmarks.find(
                (item) => item.type === 'CHARACTER' && item.id === char.id,
              );
              return (
                <Card
                  key={char.id}
                  id={char.id}
                  name={char.name}
                  description={char.description ?? ''}
                  bookmarkClassName={
                    inBookmark
                      ? 'btn btn-action btn-bookmark bookmark-selected'
                      : 'btn btn-action btn-bookmark'
                  }
                  bookmarkIcon={inBookmark ? faBookmarkSolid : faBookmarkRegular}
                  handleViewMore={handleViewMore}
                  handleHideItem={handleHideItem}
                  handleBookmarkAction={inBookmark ? handleRemoveBookmark : handleAddBookmark}
                  imageUrl={
                    char.thumbnail
                      ? `${char.thumbnail.path}/portrait_uncanny.${char.thumbnail.extension}`
                      : imagePlaceHolder
                  }
                />
              );
            })}
        </div>
      </div>
      {results && results.length <= 0 && <h2 className='error-message'>Characters not found üòÆ</h2>}
      {!loading && results && results.length > 0 && characters && characters.length <= 0 && (
        <h2 className='error-message'>Characters on this page are hidden ü§ê</h2>
      )}
      {!loading && !error && results && genericResponse && results.length > 0 && (
        <Pagination genericResponse={genericResponse} onChange={handleChangePage} />
      )}
    </div>
  );
};

export default ListCharactersPage;
