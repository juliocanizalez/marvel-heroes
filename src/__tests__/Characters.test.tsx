/* eslint-disable arrow-body-style */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-unused-vars */
/* eslint-disable import/extensions */
/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import configureStore from 'redux-mock-store';

import store from '../store/store';
import { setName } from '../actions/Search';
import { addBookmark } from '../actions/LocalItems';
import server from '../mocks/server';
import { emptyResponse } from '../mocks/testData';
import Characters from '../containers/Characters/Characters';

describe('Test on Characters component', () => {
  test('Should render component correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
      expect(container).toMatchSnapshot();
    });
  });

  test('should show typed name on input', async () => {
    const search = 'spid';
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    userEvent.type(screen.getByPlaceholderText(/Hero´s name/i), search);
    expect((screen.getByPlaceholderText(/Hero´s name/i) as HTMLInputElement).value).toBe(search);
  });

  test('should add/remove bookmark', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
    });

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).not.toBe(null);

    userEvent.click(container.querySelector('.btn-bookmark') as Element);

    expect(container.querySelector('.bookmark-selected')).toBe(null);
  });

  test('should hide card and show hidden characters message', async () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.cards')).not.toBeNull();
    });

    userEvent.click(container.querySelector('.btn-hide') as Element);

    expect(container.querySelector('.card')).toBeNull();
    expect(screen.getByText(/Characters on this page are hidden 🤐/i)).toBeInTheDocument();
  });

  test('should handle no results', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters`, (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(emptyResponse()));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Characters not found 😮/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
    });
  });

  test('should handle api error', async () => {
    server.use(
      rest.get(`${process.env.REACT_APP_API_URL}v1/public/characters`, (req, res, ctx) => {
        return res(ctx.status(404));
      }),
    );

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(screen.getByText(/Could not load characters 😓/i)).toBeInTheDocument();
      expect(container.querySelector('.cards')).toBeNull();
    });
  });
});

describe('Mock store tests', () => {
  const mockStore = configureStore();

  const initialState = {
    search: {
      offset: 0,
      limit: 12,
      baseUrl: '',
      url: '',
      name: '',
      comic: '',
      story: '',
      loading: false,
      error: '',
      data: {
        data: {
          offset: 0,
          limit: 12,
          total: 1,
          count: 1,
          results: [
            {
              id: 1,
              name: 'spiderman',
              description: 'description',
              thumbnail: {
                path: 'path',
                extension: 'jpg',
              },
            },
          ],
        },
      },
    },
    localItems: {
      hiddenItems: [],
      bookmarks: [],
    },
  };
  const testStore = mockStore(initialState);

  testStore.dispatch = jest.fn();

  jest.mock('../actions/Search', () => ({
    setBaseUrl: jest.fn(),
    setName: jest.fn(),
    setAllParams: jest.fn(),
  }));

  beforeEach(() => {
    testStore.clearActions();
    jest.clearAllMocks();
  });

  test('should call set name dispatch while typing name', async () => {
    const searchValue = 'spid';
    render(
      <Provider store={testStore}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    userEvent.type(screen.getByPlaceholderText(/Hero´s name/i), searchValue);

    const searchValueArray = searchValue.split('');
    // eslint-disable-next-line max-len
    searchValueArray.forEach((letter) =>
      expect(testStore.dispatch).toHaveBeenCalledWith(setName(letter)),
    );
  });

  test('should dispatch add bookmark', () => {
    const { container } = render(
      <Provider store={testStore}>
        <MemoryRouter>
          <Characters />
        </MemoryRouter>
      </Provider>,
    );

    expect(container.querySelector('.cards')).not.toBeNull();
    userEvent.click(container.querySelector('.btn-bookmark') as Element);
    expect(testStore.dispatch).toHaveBeenCalledWith(addBookmark({ id: 1, type: 'CHARACTER' }));
  });
});
