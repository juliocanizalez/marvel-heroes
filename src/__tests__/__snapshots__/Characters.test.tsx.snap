// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Characters component Should render component correctly 1`] = `
<div>
  <div
    class="main-content "
  >
    <div
      class="page-title"
    >
      <h1>
        Characters
      </h1>
    </div>
    <div
      class="search-filters-form mb-5"
    >
      <form>
        <div
          class="search-title"
        >
          Search your hero
        </div>
        <div
          class="search-header__list"
        >
          <h3>
            Name
          </h3>
          <h3>
            Comic
          </h3>
          <h3>
            Story
          </h3>
        </div>
        <div
          class="search-select"
        >
          <div
            class="search-value"
          >
            <input
              autocomplete="off"
              class="search-input"
              name="name"
              placeholder="HeroÂ´s name"
              type="text"
              value=""
            />
          </div>
          <div
            class="search-value"
          >
            <select
              class="search-select"
            >
              <option
                value=""
              >
                Select a comic
              </option>
            </select>
          </div>
          <div
            class="search-value"
          >
            <select
              class="search-select"
            >
              <option
                value=""
              >
                Select a story
              </option>
            </select>
          </div>
        </div>
      </form>
    </div>
    <div
      class="loading"
    >
      <div
        class="lds-dual-ring"
      />
    </div>
    
    <div
      class="cards"
    >
      <div
        class="cards-content"
      />
    </div>
  </div>
</div>
`;
